src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,488 @@
+﻿/* Copyright 2020–present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Security;
+using System.Threading.Tasks;
+using MongoDB.Bson;
+using MongoDB.Bson.IO;
+using MongoDB.Bson.Serialization;
+using MongoDB.Bson.Serialization.Serializers;
+using MongoDB.Driver.Core.Connections;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.Core.Authentication
+{
+    /// <summary>
+    /// The Mongo AWS authenticator.
+    /// </summary>
+    public class MongoAWSAuthenticator : SaslAuthenticator
+    {
+        private static readonly Lazy<HttpClient> _httpClientInstance = new Lazy<HttpClient>(() =>
+        {
+            var httpClient = new HttpClient();
+            httpClient.Timeout = TimeSpan.FromSeconds(10);
+            return httpClient;
+        });
+
+        // constants
+        private const string awsContainerCredentialsRelativeUriKey = "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI";
+        private const string awsSessionTokenMechanismPropertyKey = "AWS_SESSION_TOKEN";
+        private const string ecsBaseUri = "http://169.254.170.2";
+        private const string ec2BaseUri = "http://169.254.169.254";
+        private const int randomLength = 32;
```

Comment:
Constant values taken from spec: [link](https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#obtaining-credentials)

Reply:
Mostly we use the upper camelcase for constants names

Reply:
Also, the constant name does tell for what reason this value will be used. I think this value may be moved into the method `private byte[] GenerateRandomBytes()` similar how it's done in  https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Authentication/ScramShaAuthenticator.cs#L160

Reply:
It's also used in server nonce validation here: https://github.com/MikalaiMazurenka/mongo-csharp-driver/pull/22/files#diff-ffb35b98eecd05d3c7b6133327bf13b3R436
So to avoid using magic numbers in code, I leave it as is. Regarding case: I see different formats throughout the solution. E.g. pascal case: [link](https://github.com/mongodb/mongo-csharp-driver/blob/719731cb7548ae6c1cfc253c33384084f84fb3b1/src/MongoDB.Driver.Core/Core/Authentication/SaslPrepHelper.cs#L45), winapi style: [link](https://github.com/mongodb/mongo-csharp-driver/blob/719731cb7548ae6c1cfc253c33384084f84fb3b1/src/MongoDB.Driver.Core/Core/Authentication/Sspi/NativeMethods.cs#L27) (not sure how it's called), camel case: [link](https://github.com/mongodb/mongo-csharp-driver/blob/719731cb7548ae6c1cfc253c33384084f84fb3b1/src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs#L56). Could you please link some guidelines on this?
For now I've made the name more meaningful.

Reply:
The only guide is to use the most latest way, and as far as I know, in the latest code, we use the upper camelcase format. It's a good point to mention it somewhere, I think we have some document which can be used for that target. @vincentkam do you remember the document which we wanted to use for that target (I remember we discussed similar case around a year ago)?